### ISSUE TITLE
float4x4 inverse producing nans on ARM64

### ISSUE THREAD

**Issue by lordalcol** [t0rakka/mango#31]
Thanks for the great work!

I am having an issue with the `inverse` function running on ARMv8. The following code:

```
auto m = mango::matrix::identity();
m = mango::inverse(m);
```

Produces the following values when running on an ARM 64-bit CPU (Android, compiled with clang, -O2):
```
nan, nan, nan, nan
nan, nan, nan, nan
nan, nan, nan, nan
nan, nan, nan, nan
```

While on x86 (MSVC) I get:

```
1.000, -0.000, -0.000, -0.000
-0.000, 1.000, -0.000, -0.000
0.000, 0.000, 1.000, -0.000
0.000, 0.000, -0.000, 1.000
```

**Comment by t0rakka** [comment]
math.cpp, line 29 is invoking this code:

        m[0] = simd::set_component<0>(zero, s);
        m[1] = simd::set_component<1>(zero, s);
        m[2] = simd::set_component<2>(zero, s);
        m[3] = simd::set_component<3>(zero, 1.0f);

The problem should be in ARM/NEON set_component.. actually, above is not the most efficient way to write the scaling matrix setup. I will fix both, but need my ARM hardware so allow a bit headroom for testing.

**Comment by t0rakka** [comment]
Nothing immediately obviously wrong in the ARM code:

    template <unsigned int Index>
    static inline f32x4 set_component(f32x4 a, f32 s)
    {
        static_assert(Index < 4, "Index out of range.");
        return vsetq_lane_f32(s, a, Index);
    }

Looks pretty good. Can't find obvious bug from looking at the code. Will look again with the ARM board to see which step goofs this up. If it works on my super-convenient ARM cross compiler environment then it could be a compiler bug.

You mention Android; what's the NDK version you are using? I want to make sure I am as close to your environment as possible in case this becomes hard-to-reproduce in my default setup.

Of course, it could be in the inverse as well but the first thing is to validate that the input is good (ie. the matrix we are feeding is good data).

**Comment by t0rakka** [comment]
Tested on:

uname -a

"Linux odroid 3.8.13.30 #1 SMP PREEMPT Fri Sep 4 23:45:57 BRT 2015 armv7l armv7l armv7l GNU/Linux

Compiler:

g++ (Ubuntu/Linaro 7.4.0-1ubuntu1~14.04~ppa1) 7.4.0

mango::getSystemInfo():

Platform: "Linux"
CPU: "ARM", Bits: 32, Endian: LITTLE 
Detected CPU Features: N/A
Compiled SIMD Features: NEON 
Hardware threads: 4
Build: Mar 25 2019  16:41:00

1.000000 0.000000 -0.000000 -0.000000
0.000000 1.000000 -0.000000 -0.000000
0.000000 0.000000 1.000000 -0.000000
0.000000 0.000000 -0.000000 1.000000

The 32 bit ARM/NEON path seems to be OK with above setup so there is no fundamental flaw in ARM build. Could be clang or ARM64 specific quirk.. trying to repro again..

**Comment by lordalcol** [comment]
Thank you for checking it. I can test it for you, if you tell me what to look out for. (i.e. some intermediate value?)

**Comment by t0rakka** [comment]
If you could do that it would help reproducing. 

I'm pretty sure it's something I have done, so far my best bet is the division; it is different on ARMv7 and ARMv8, the 7 can only do reciprocal and the 8 can do a "real" division:

#ifdef __aarch64__
    static inline f32x4 div(f32x4 a, f32x4 b)
    {
        return vdivq_f32(a, b);
    }

    static inline f32x4 div(f32x4 a, f32 b)
    {
        f32x4 s = vdupq_n_f32(b);
        return vdivq_f32(a, s);
    }
#else
    static inline f32x4 div(f32x4 a, f32x4 b)
    {
        f32x4 n = vrecpeq_f32(b);
        n = vmulq_f32(vrecpsq_f32(n, b), n);
        n = vmulq_f32(vrecpsq_f32(n, b), n);
        return vmulq_f32(a, n);
    }

    static inline f32x4 div(f32x4 a, f32 b)
    {
        f32x4 s = vdupq_n_f32(b);
        f32x4 n = vrecpeq_f32(s);
        n = vmulq_f32(vrecpsq_f32(n, s), n);
        n = vmulq_f32(vrecpsq_f32(n, s), n);
        return vmulq_f32(a, n);
    }
#endif

That's the one difference there is between v7 and v8 that is used from the inverse function. You could force the reciprocal variant on v8 and see what that does.

**Comment by t0rakka** [comment]
Progress update. Debugging on ARM64, getting Inf/NaN's the "denom" is (0, 0, 0, 0) before the division. Pin-pointing why, current suspect: shuffles on ARM64.

**Comment by t0rakka** [comment]
Found it, fantastic; the __builtin_shufflevector DOESN'T WORK CORRECTLY. I disable this "optimization" for now.

    template <u32 x, u32 y, u32 z, u32 w>
    inline f32x4 shuffle(f32x4 a, f32x4 b)
    {
        static_assert(x < 4 && y < 4 && z < 4 && w < 4, "Index out of range.");
        return __builtin_shufflevector(a.data, b.data, x, y, z, w);
    }

    template <u32 x, u32 y, u32 z, u32 w>
    inline f32x4 shuffle(f32x4 v)
    {
        static_assert(x < 4 && y < 4 && z < 4 && w < 4, "Index out of range.");
        return __builtin_shufflevector(v.data, v.data, x, y, z, w);
    }

**Comment by t0rakka** [comment]
Should be fixed in https://github.com/t0rakka/mango/commit/b00466583472f7c96faee3cb1405b6e24b1b02d2

Feel free to close the issue after you verify that the problem is fixed. Thanks for the report. :)

**Comment by lordalcol** [comment]
Wow, this was over my head. Thank you!
I can confirm that the bug has been fixed in b004655.
